core/src/main/java/com/datastax/oss/driver/internal/core/control/ControlConnection.java

```diff
@@ -294,6 +293,15 @@ private void init(boolean listenToClusterEvents, boolean reconnectOnFailure) {
       }
     }
 
+    private ReconnectionSchedule newSchedule() {
+      boolean isInitialConnection =
+          reconnectOnInit
+              // This is always set after the initial attempt, regardless of its outcome.
+              // If the first attempt failed, we create the schedule before setting it.
+              && !firstConnectionAttemptFuture.isDone();
```

Comment:
This is overly complex to follow, couldn't you create a dedicated boolean variable to store that information (initial/not initial) instead?

Besides, are you sure that `firstConnectionAttemptFuture.isDone()` is always false for the initial attempt? From what I understand, in case the initial attempt fails, the thread executing the initial sequence will first call `reconnection.start()` line 276 then complete `firstConnectionAttemptFuture` exceptionally line 289; but since the reconnection process is async, the future may or may not be completed before `newSchedule` is called.

Reply:
I don't think concurrency is an issue, `Reconnection` uses the same admin thread and `Reconnection.start()` creates the schedule synchronously.

But you are right that the code is too convoluted. What makes it complicated is that the control connection can be initialized in two ways:

1. by default, it's the component that determines if the initial connection to the cluster succeeds.
2. But you could also have a custom `TopologyMonitor` that does that a different way, and only use the control connection for schema metadata and server events. In that case the control connection is initialized late, and `newControlConnectionSchedule()` must never be called with a `true` argument.

I think I can make the code more clear if there are separate init methods for those two cases. I'll give it a try.

Reply:
> I don't think concurrency is an issue, Reconnection uses the same admin thread and Reconnection.start() creates the schedule synchronously.

Hmm my bad, `newSchedule()` is invoked synchronously while `reconnect()` is invoked asynchronously, so the future will never be done when `newSchedule()` is executed.

> I think I can make the code more clear if there are separate init methods for those two cases. I'll give it a try.

💯